soql-standalone-parser/src/main/scala/com/socrata/soql/BinaryTree.scala

```diff
@@ -0,0 +1,90 @@
+package com.socrata.soql
+
+trait BinaryTree[+T] {
```

Comment:
This should be `sealed` so that the compiler can do exhaustivity checking (and I know it's extended outside this file, and also that there are matches in here which are non-exhaustive as a result -- more specifically, I a pre-analysis tree and post-analysis tree shouldn't be the same type)

Reply:
..though parameterizing over the leaf type would deal nicely with that.

Reply:
How do I seal the trait with BinaryTree[Select] and BinaryTree[Analysis] being in a different file/package?
I don't get how case class Leaf helps.

Reply:
You do
```sealed trait BinaryTree[T] {
   ...
}
sealed trait Compound[T] extends BinaryTree[T] {
}
case class /*all the various different types of Compound */ extends Compound[T]
case class Leaf[T](leaf: T) extends BinaryTree[T]
```

Basically anything abstract gets sealed so that a pattern match can absolutely say "I know for a fact that all possible cases of the BinaryTree[T] are covered and there isn't some random `class Bleh extends BinaryTree[Select]` out in some file I don't know about which isn't being handled here."